graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/fetchers/admin/DeploySchemaFetcherBase.java

```diff
@@ -75,25 +79,41 @@
 
     String input = getSchemaContents(environment);
     UUID expectedVersion = getExpectedVersion(environment);
+    MigrationStrategy migrationStrategy = environment.getArgument("migrationStrategy");
+    boolean dryRun = environment.getArgument("dryRun");
 
+    ImmutableMap.Builder<String, Object> response = ImmutableMap.builder();
     ProcessedSchema processedSchema =
         new SchemaProcessor(authenticationService, authorizationService, dataStoreFactory)
             .process(input, keyspace);
-
-    // TODO this is racy -- figure out a way to handle concurrency better
-    SchemaSource newSource =
-        new SchemaSourceDao(dataStore).insert(namespace, expectedVersion, input);
-    processedSchema.dropAndRecreateSchema(dataStore);
-
-    // TODO update SchemaFirstCache from here instead of letting it reload from the DB
-
-    return ImmutableMap.of(
-        "version",
-        newSource.getVersion(),
+    response.put(
         "messages",
         processedSchema.getMessages().stream()
             .map(this::formatMessage)
             .collect(Collectors.toList()));
+
+    // TODO the rest of the method is racy -- we should at least ensure that two deploy mutations
+    // can't compete with each other (maybe add a "migration in progress" column in the database and
+    // do an additional LWT on it)
+    List<MigrationQuery> queries =
+        new CassandraMigrator(dataStore, processedSchema.getMappingModel(), migrationStrategy)
+            .compute();
+
+    response.put(
+        "cqlChanges",
+        queries.isEmpty()
+            ? ImmutableList.of("No changes, the CQL schema is up to date")
+            : queries.stream().map(MigrationQuery::getDescription).collect(Collectors.toList()));
+
+    if (!dryRun) {
+      queries.forEach(q -> dataStore.execute(q.getQuery()));
```

Comment:
should we use map and do something with the outcome of executed query? I.e:
```
List<ResultSet> result = List<queries.map(q -> dataStore.execute(q.getQuery());
validateIfAllApplied(result);
```

Reply:
All the CQL queries have to succeed, otherwise the deployment fails with a GraphQL error. So if you reach the point where you are reading `cqlChanges` from a `deploySchema` response, the only possible outcome is success.